numpy/f2py/crackfortran.py

```diff
@@ -2873,6 +2906,84 @@ def analyzevars(block):
 analyzeargs_re_1 = re.compile(r'\A[a-z]+[\w$]*\Z', re.I)
 
 
+def param_eval(n, varsn, v, g_params, params):
+    # Creates a dictionary of indices and values for each parameter in a
+    # parameter array to be evaluated later.
+    #
+    # WARNING: It is not possible to initialize multidimensional array
+    # parameters e.g. dimension(-3:1, 4, 3:5) at this point. This is because in
+    # Fortran initialization through array constructor requires the RESHAPE
+    # intrinsic function. Since the right-hand side of the parameter declaration
+    # is not executed in f2py, but rather at the compiled c/fortran extension, later, it is not possible
+    # to execute a reshape of a parameter array.
+    # One issue remains: if the user wants to access the array parameter from python,
+    # we should either
+    # 1) allow them to access the parameter array using python standard indexing (which
+    #    is often incompatible with the original fortran indexing)
+    # 2) allow the parameter array to be accessed in python as a dictionary with fortran
+    #    indices as keys
+    # We are choosing 2 for now.
+    if any(a[:9] == 'dimension' for a in varsn['attrspec']):
+        # This is an array parameter.
+        # First, we parse the dimension information
+        a = [item[9:] for item in varsn['attrspec'] if item[:9] == 'dimension'][0]
+        dimrange = a.lstrip('(').rstrip(')')
+        dimrange = dimrange.split(',')
+        if len(dimrange) == 1:
+            # e.g. dimension(2) or dimension(-1:1)
+            dimrange = dimrange[0].split(':')
+            # now, dimrange is a list of 1 or 2 elements
+            if len(dimrange) == 1:
+                dimrange = range(1, int(dimrange[0])+1)
+            else:
+                dimrange = range(int(dimrange[0]), int(dimrange[1])+1)
+        else:
+            print(n)
+            raise Exception('param_eval: multidimensional array parameters not supported: %s\n' % repr(n))
+
+        # Parse parameter value
+        v = v.lstrip('(/').rstrip('/)').split(',')
+        v_eval = []
+        for item in v:
+            try:
+                item = eval(item, g_params, params)
+            except Exception as msg:
+                v_eval.append(item)
+                outmess('param_eval: got "%s" on %s\n' % (msg, repr(item)))
+            v_eval.append(item)
+
+        p = dict([])
+        for i in range(len(dimrange)):
+            p[dimrange[i]] = v_eval[i]
+    else:
+        try:
+            p = eval(v, g_params, params)
+        except Exception as msg:
+            p = v
+            outmess('param_eval: got "%s" on %s\n' % (msg, repr(v)))
+    return p
+
+
+def param_parse(d, params):
```

Comment:
Could you add a docstring with two examples showing what this function does in the recursive and non-recursive case?